src/main/java/com/internet/shop/dao/jdbc/UserDaoJdbcImpl.java

```diff
@@ -0,0 +1,200 @@
+import com.internet.shop.model.User;
+import com.internet.shop.util.ConnectionUtil;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+@Dao
+public class UserDaoJdbcImpl implements UserDao {
+    @Override
+    public Optional<User> findByLogin(String login) {
+        User user = new User();
+        String query = "SELECT * FROM users WHERE deleted = false AND login = ?;";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setString(1, login);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                user = getUserFromResultSet(resultSet);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can not get user from DB with login = " + login, e);
+        }
+        user.setRoles(getRolesOfUser(user.getId()));
+        return Optional.ofNullable(user);
+    }
+
+    @Override
+    public User create(User item) {
+        String query = "INSERT INTO users(name, login, password) VALUES(?, ?, ?)";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query,
+                        Statement.RETURN_GENERATED_KEYS)) {
+            statement.setString(1, item.getName());
+            statement.setString(2, item.getLogin());
+            statement.setString(3, item.getPassword());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                item.setId(resultSet.getLong(1));
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Unable to create user " + item, e);
+        }
+        return defineRoles(item);
+    }
+
+    @Override
+    public Optional<User> getById(Long item) {
+        User user = new User();
+        String query = "SELECT * FROM users "
+                + "INNER JOIN users_roles ON users.user_id = users_roles.user_id "
+                + "WHERE users.deleted = false AND users.user_id = ?;";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement statement = connection.prepareStatement(query)) {
+            statement.setLong(1, item);
+            ResultSet resultSet = statement.executeQuery();
+            if (resultSet.next()) {
+                user = getUserFromResultSet(resultSet);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Unable to create user with ID " + item, e);
+        }
+        user.setRoles(getRolesOfUser(user.getId()));
```

Comment:
What about defineRoles method?

Reply:
defineRoles is only called when we call the methods "create" and "update".